src/test/function_test/test_backup_and_restore.cpp

```diff
@@ -0,0 +1,247 @@
+        bool is_backup_complete = false;
+        while (!is_backup_complete && sleep_sec <= max_sleep_seconds) {
+            std::cout << "sleep 10s to wait backup complete." << std::endl;
+            sleep(10);
+            sleep_sec += 10;
+
+            auto resp = query_backup(backup_id);
+            if (resp.err != ERR_OK) {
+                return false;
+            }
+            // we got only one backup_item for a certain app_id and backup_id.
+            auto item = resp.backup_items[0];
+            is_backup_complete = (item.end_time_ms > 0);
+        }
+        return is_backup_complete;
+    }
+
+    bool wait_app_become_healthy(const std::string &app_name, int max_sleep_seconds)
+    {
+        int sleep_sec = 0;
+        bool is_app_healthy = false;
+        while (!is_app_healthy && sleep_sec <= max_sleep_seconds) {
+            std::cout << "sleep 10s to wait app become healthy." << std::endl;
+            sleep(10);
+            sleep_sec += 10;
+
+            int32_t new_app_id;
+            int32_t partition_count;
+            std::vector<partition_configuration> partitions;
+            auto err = _ddl_client->list_app(app_name, _old_app_id, partition_count, partitions);
+            if (err != ERR_OK) {
+                std::cout << "list app " + app_name + " failed" << std::endl;
+                return false;
+            }
+            int32_t healthy_partition_count = 0;
+            for (const auto &partition : partitions) {
+                if (partition.primary.is_invalid()) {
+                    break;
+                }
+                if (partition.secondaries.size() + 1 < partition.max_replica_count) {
+                    break;
+                }
+                healthy_partition_count++;
+            }
+            is_app_healthy = (healthy_partition_count == partition_count);
+        }
+        return is_app_healthy;
+    }
+
+    void test_backup_and_restore(const std::string &user_specified_path = "")
+    {
+        error_code err = _ddl_client->create_app(_old_app_name, "pegasus", 4, 3, {}, false);
+        ASSERT_EQ(ERR_OK, err);
+        ASSERT_TRUE(wait_app_become_healthy(_old_app_name, 180));
+
+        ASSERT_TRUE(write_data());
+        ASSERT_TRUE(verify_data(_old_app_name));
+
+        auto resp = start_backup(user_specified_path);
+        ASSERT_EQ(ERR_OK, resp.err);
+        int64_t backup_id = resp.backup_id;
+        ASSERT_TRUE(wait_backup_complete(backup_id, 180));
+        err = start_restore(backup_id, user_specified_path);
+        ASSERT_EQ(ERR_OK, err);
+        ASSERT_TRUE(wait_app_become_healthy(_new_app_name, 180));
+
+        ASSERT_TRUE(verify_data(_new_app_name));
+    }
+
+private:
+    std::shared_ptr<replication_ddl_client> _ddl_client;
+
+    const int _num_of_rows;
+    const std::string _cluster_name;
+    const std::string _old_app_name;
+    const std::string _new_app_name;
+    const std::string _provider;
+
+    std::string _pegasus_root_dir;
```

Comment:
dsn::string_view _pegasus_root_dir;

Reply:
`global_env::instance()._pegasus_root` is a `string`, so maybe no need to convert it to `dsn::string_view`.